src/services/suggestionDiagnostics.ts

```diff
@@ -104,4 +122,34 @@ namespace ts {
     function getErrorNodeFromCommonJsIndicator(commonJsModuleIndicator: Node): Node {
         return isBinaryExpression(commonJsModuleIndicator) ? commonJsModuleIndicator.left : commonJsModuleIndicator;
     }
+
+    function isPromiseType(T: Type): boolean {
+        return T.flags === TypeFlags.Object && T.symbol.name === "Promise";
+    }
+
+    function getReturnStmts(node: Node, retStmts: ReturnStatement[]) {
+        forEachChild(node, visit);
+
+        function visit(node: Node) {
+            if (isFunctionLike(node)) {
+                return;
+            }
+
+            if (node.kind === SyntaxKind.ReturnStatement) {
+                forEachChild(node, hasCallback);

```

Comment:
you do not need the call to `forEachChild` here. you can just check if the expression hasCallback. e.g.

```ts
function visit(node: Node) {
    if (isReturnStatement(node)) {
        if (isCallExpression(child) && isPropertyAccessExpression(child.expression) &&
            (child.expression.name.text === "then" || child.expression.name.text === "catch" || child.expression.name.text === "finally")) {
            result.push(node);
        }
    }
    else if (!isFunctionLike(node)) {
        forEachChild(node, visit);
    }
}
```

Reply:
Also using the functions `isCallExpression` and `isPropertyAccessExpression` avoids the need for casts.

Reply:
I think we do still need to visit all children. The node we are passing in is the entire function declaration.